---
title: "Lab10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Libraries
```{r}
library(vcfR)
library(adegenet)
library(reshape2)
library(ggplot2)
library(knitr)
```

# VCF files
```{r}
data(vcfR_example)
vcf
```

# The meta region
```{r}
strwrap(vcf@meta[1:7])

queryMETA(vcf)
queryMETA(vcf, element = 'DP')
queryMETA(vcf, element = 'FORMAT=<ID=DP')
```

# The fix region
```{r}
head(getFIX(vcf))
```

# The gt region
```{r}
vcf@gt[1:6, 1:4]
```

# vcfR
```{r}
vcf <- read.vcfR("https://github.com/grunwaldlab/Population_Genetics_in_R/blob/master/pinfsc50_filtered.vcf.gz")

head(vcf)
write.vcf(vcf, "myVCFdata_filtered.vcf.gz")
```

# Exercises Part 1
1. How would we find more information about read.vcfR()?
```{r}
?read.vcfR
```

2. How would we learn what the acronym “AD” stands for?
```{r}
queryMETA(vcf, element = 'AD')
```

3. We used the head() function to view the first few lines of fix data. How would we view the last few lines of fix data?
```{r}
tail(vcf@fix)
```

4. There is a column in the fix portion of the data called QUAL. It is not defined in the meta portion of the data because it is defined in the VCF specification. It stands for ‘quality’. Does QUAL appear useful to us? Why or why not?
```{r}
plot(vcf)

# Yes it is important, QUAL shows the quality of the genes and can be the first step in allowing you to filter high and low quality genes.
```

5. How would we query the sample names?
```{r}
colnames(vcf@gt)
```

# Opening and examining the dataset
```{r}
vcf <- read.vcfR("https://github.com/grunwaldlab/Population_Genetics_in_R/blob/master/pinfsc50_filtered.vcf.gz")
vcf
```

# Converting VCF data to a genlight object
```{r}
x <- vcfR2genlight(vcf)
x

gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:3]

t(as.matrix(x))[c(1,5,17), 1:3]

pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x)

ploidy(x) <- 2
```

# Distance matrices
```{r}
x.dist <- dist(x)

x.dist <- poppr::bitwise.dist(x)
```

# Creating chromR objects
```{r}
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")

chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)

chrom

plot(chrom)
chromoqc(chrom, dp.alpha = 66)

chrom <- proc.chromR(chrom, verbose = TRUE)
plot(chrom)
chromoqc(chrom, dp.alpha = 66)

vcf <- read.vcfR("pinfsc50_filtered.vcf.gz", verbose = FALSE)
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66)
```

# Tabular summaries
```{r}
head(chrom@var.info)
head(chrom@win.info)
```

# Genetic differentiation
```{r}
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))

knitr::kable(head(myDiff[,16:19]))

knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))

dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)

p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

# Exercises Part 2
1. You actually have everything you need to make a Manhattan plot. Can you figure out how to plot G′ST (y-axis) by genomic position (POS)?
```{r}
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))
```

2. This Manhatttan plot shouldlook a bit unusual. Can you think of anything that may be wrong with this analysis?
```{r}
table(pop)
# The sample size is too small to give an accurate read
```

3. Can you figure out how to zoom in on a particular region of a chromosome in chromoqc()?
```{r}
chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```

4. Can you use the function queryMETA() to look for other data in your file that may be of interest?
```{r}
queryMETA(vcf)
```



